File: general-interview-prep/javascript/boundaryOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var boundaryOfBinaryTree = function(root) {
    if(!root || root == null) return [];
    if(!root.left && !root.right) return [root.val];

    let resL = [],
        resR = [],
        res = [];

    //DFS push all the nodes on the left side except root and the leaves
    function boundaryL(root){
        if(!root) return;
        if(root.left){
            resL.push(root.val);
            boundaryL(root.left);
        }
        else if(root.right){
            resL.push(root.val);
            boundaryL(root.right);
        }

    }

	//check the leaves.  if without any left and right nodes, push to res
    function leaf(root){
        if(!root) return;
        leaf(root.left);
        if(!root.left && !root.right){
            res.push(root.val);
        }
        leaf(root.right);
    }

     //DFS push all the nodes on the right side except root and the leaves.
    function boundaryR(root){
        if(!root) return;
        if(root.right){
            boundaryR(root.right); // do the recursion first to get the root.val bottom up
            resR.push(root.val);
        }
        else if(root.left){
            boundaryR(root.left);
            resR.push(root.val);
        }
    }

    boundaryL(root.left);
    leaf(root.left);
    leaf(root.right);
    boundaryR(root.right);
    return [root.val,...resL,...res,...resR];

};




/*
 *
 * Even Easier implementation with recursion
 *
 */
 function buildSide(root, side, res) {
     if (!root) return;
     if (root.left || root.right) res.push(root.val);
     if (root[side]) buildSide(root[side], side, res);
     else buildSide(root[side === 'left' ? 'right' : 'left'], side, res);
 }

 function buildBottom(root, res) {
     if (!root) return;
     if (!root.left && !root.right) res.push(root.val);
     buildBottom(root.left, res);
     buildBottom(root.right, res);
 }

 const simpleBoundaryOfBinaryTree = function(root) {
     if (!root) return [];
     if (!root.left && !root.right) return [root.val];
     const res = [root.val];
     buildSide(root.left, 'left', res);
     buildBottom(root, res);
     const right = [];
     buildSide(root.right, 'right', right);
     return [...res, ...right.reverse()];
 };

File: general-interview-prep/golang/boundaryOfBinaryTree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func buildBottom(node *TreeNode, stack []int) []int {
    if node == nil {return stack}
    if node.Left == nil && node.Right == nil {
        stack = append(stack, node.Val)
    }

    stack = buildBottom(node.Left, stack)
    stack = buildBottom(node.Right, stack)
    return stack
}

func buildSide(node *TreeNode, side string, stack []int) []int {
    if node == nil {return stack}
    if node.Left != nil || node.Right != nil {
        stack = append(stack, node.Val)
    }
    if side == "left" {
        if node.Left != nil {
            stack = buildSide(node.Left, "left", stack)
        } else {
            stack = buildSide(node.Right, "left", stack)
        }
    } else {
        if node.Right != nil {
            stack = buildSide(node.Right, "right", stack)
        } else {
            stack = buildSide(node.Left, "right", stack)
        }
    }

    return stack
}

func reverseSlice(arr []int) []int {
    temp := []int{}

    for len(arr) != 0 {
        temp = append(temp, arr[len(arr) - 1])
        arr = arr[:len(arr) - 1]
    }

    return temp
}

func boundaryOfBinaryTree(root *TreeNode) []int {
    if root == nil {return []int{}}
    if root.Left == nil && root.Right == nil {
        return []int{root.Val}
    }

    res := []int{root.Val}
    res = buildSide(root.Left, "left", res)
    res = buildBottom(root, res)

    right := []int{}
    right = buildSide(root.Right, "right", right)

    result := []int{}
    result = append(result, res...)
    result = append(result, reverseSlice(right)...)
    return result

}
